### DISCUSSION TITLE
Zoom to elements widget

### DISCUSSION THREAD

**Discussion by brinpat** [iTwin/community#7] [General]
When clicking on an asset and trying to use Zoom to Elements widget, nothing happens. This feature was working but seems to no longer work. 

I am expecting to be able to change the view using a selected asset as a view pivot.

**Comment by pmconne** [comment]
Hi @brinpat.
Can you please specify the version numbers of the packages you upgraded to and the version numbers you upgraded from?

**Reply by brinpat** [reply]
Hi @pmconne   

I am unsure what versions I upgraded from as it was a while since I upgraded, but these are the current version numbers:

  "@bentley/icons-generic": "^1.0.34",
    "@bentley/react-scripts": "^5.0.3",
    "@itwin/appui-abstract": "^3.5.4",
    "@itwin/appui-layout-react": "^3.5.4",
    "@itwin/appui-react": "^3.5.4",
    "@itwin/browser-authorization": "^0.5.1",
    "@itwin/clash-detection-client": "^0.2.1",
    "@itwin/components-react": "^3.5.4",
    "@itwin/core-bentley": "^3.5.4",
    "@itwin/core-common": "^3.5.4",
    "@itwin/core-frontend": "^3.5.4",
    "@itwin/core-geometry": "^3.5.4",
    "@itwin/core-i18n": "^3.5.4",
    "@itwin/core-markup": "^3.5.4",
    "@itwin/core-orbitgt": "^3.5.4",
    "@itwin/core-quantity": "^3.5.4",
    "@itwin/core-react": "^3.5.4",
    "@itwin/core-telemetry": "^3.5.4",
    "@itwin/imodel-components-react": "^3.5.4",
    "@itwin/imodels-access-frontend": "^1.0.2",
    "@itwin/itwinui-react": "^1.48.1",
    "@itwin/measure-tools-react": "^0.7.3",
    "@itwin/presentation-common": "^3.5.4",
    "@itwin/presentation-frontend": "^3.5.4",
    "@itwin/property-grid-react": "^0.5.5",
    "@itwin/tree-widget-react": "^0.4.7",
    "@itwin/web-viewer-react": "^3.1.2",
    "@itwin/webgl-compatibility": "^3.5.4",
    "@types/history": "^4.7.11",
    "@types/node": "^14.18.36",
    "@types/react": "^17.0.53",
    "@types/react-dom": "^17.0.18",
    "history": "^4.10.1",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-redux": "^7.2.9",
    "redux": "^4.2.0",
    "typescript": "~4.4.0",
    "yarn": "^1.22.19"

**Reply by pmconne** [reply]
Thanks. As far as I can tell, the underlying `Viewport.zoomTo*` APIs haven't changed since 3.1.0.
@raplemie are you aware of any regressions with this widget?

**Reply by raplemie** [reply]
Hi @brinpat, can you send a screencap of where you start this ? We dont provide a "Zoom to Element" widget at the AppUI library level, so this may come from one of the other package you use also provided by iTwin, with a screencap we could probably figure proper help.

(Remember that this is a public repository, so you might want to hide private information)

**Comment by brinpat** [comment]
Hi @raplemie it comes from: https://www.itwinjs.org/sandboxes/iTwinPlatform/Zoom%20to%20Elements

I believe my code for the widget is the same, though note the code for the app will be different.

**Comment by pmconne** [comment]
Looks like an assembly scope problem.
The sandbox widget sets scope to select the top assembly:
```
  useEffect(() => {
    // Subscribe for unified selection changes
    // Change the default selection scope. Top-assembly scope returns key of selected element's topmost parent element (or just the element if it has no parents)
    Presentation.selection.scopes.activeScope = "top-assembly";
    Presentation.selection.selectionChange.addListener(_onSelectionChanged);
  }, []);
```

Select this big blue element and you can successfully zoom to it, because it's not an assembly:
![image](https://github.com/iTwin/community/assets/22944042/6ec7de8a-ced9-4efa-9303-da3c343153f1)

Select one of the little elements coming out of the top of the blue element:
![image](https://github.com/iTwin/community/assets/22944042/629490e1-353e-48de-bbc3-06e2ef358f60)

Note that all of those elements become hilited when you click on one of them, because they all belong to the same assembly. The assembly element itself has no geometry, therefore no bounding box, therefore nothing to zoom to.

@Ellord207 did anything about this sample change when upgrading it? Did it really used to work using top-assembly scope to select assemblies with no bounding box?
@grigasp @saskliutas did something change with presentation selection scope to explain this behavior?

**Reply by grigasp** [reply]
We didn't change anything. I believe the sample should get hilite set from the selected elements and zoom to elements in that set. See `applyZoom` in [my fork](https://www.itwinjs.org/sandboxes/grigas/zoom-to-elements-fix).

**Reply by pmconne** [reply]
Thanks @grigasp - that was my expectation.
I think the same result be achieved synchronously by zooming to [iModel.hilited.elements](https://www.itwinjs.org/reference/core-frontend/selectionset/hiliteset/), which would have been populated by presentation already, except this sample has this weird extra step of requiring the user to manually add selected elements to a list instead of just using the selection set directly.

@brinpat, when comparing the older (working) version of your code with the new (non-working) version, were you selecting the same elements both times?
The results will differ based on whether the element is an assembly or not and what selection scope you are using.

**Reply by brinpat** [reply]
@pmconne the older working code allowed me to select different elements and change the view point. It would also show the element I had selected in the selected elements table. 
![image](https://github.com/iTwin/community/assets/94605936/cf3c1183-236a-4d7f-92c5-82f13d1833b3)
With the new code it does not pick up the selected element in the table either.

**Reply by Ellord207** [reply]
> @Ellord207 did anything about this sample change when upgrading it? Did it really used to work using top-assembly scope to select assemblies with no bounding box?

In the upgrade to 4.x, nothing related to this issue changed.  I suspect it was a regression before that.  I will see that the sample gets updated based on this conversation.

**Reply by brinpat** [reply]
@Ellord207 any update with this?

**Reply by raplemie** [reply]
@brinpat Opening the linked sandbox now shows that it currently uses 4.x packages, and appear to be working correctly, can you validate if the new code fixes your issue ?

**Reply by brinpat** [reply]
It is not working yet, but I noticed that in order to make the code work I have a slight difference in the imports.

I use the following:

```
import { AbstractWidgetProps, StagePanelLocation, StagePanelSection, UiItemsProvider, WidgetState } from "@itwin/appui-abstract";
import { useActiveIModelConnection } from "@itwin/appui-react";
```


While the sandbox uses:

```
import {
  StagePanelLocation,
  StagePanelSection,
  UiItemsProvider,
  useActiveIModelConnection,
  Widget,
  WidgetState
} from "@itwin/appui-react";
```

**Reply by brinpat** [reply]
I tried using the sandbox import but it comes up with errors

**Reply by Ellord207** [reply]
> @Ellord207 any update with this?

No actions have been taken as it sounded like suggested solution by @grigasp, "See applyZoom in [my fork](https://www.itwinjs.org/sandboxes/grigas/zoom-to-elements-fix)," didn't solve your problem.

> With the new code it does not pick up the selected element in the table either.

I was waiting to for the discussion to continue.  Sorry if there was a miss-understanding on my part.

**Reply by Ellord207** [reply]
> I tried using the sandbox import but it comes up with errors

The packages you mentioned are related to the reorganization of the UI packages during the upgrade to iTiwn.js v4 and should have no impact on this topic.

**Comment by brinpat** [comment]
Thanks all and thanks @Ellord207 this has been fixed once using the sandbox. After updating the code I was not adding the asset element into the table which was the issue.
